Utils/Dataflow/069_upload2es/consistency.py

```diff
@@ -0,0 +1,211 @@
+    :param fields: field names
+    :type fields: list
+
+    :return: field values, or False if the document was not found
+    :rtype: dict or bool
+    '''
+    try:
+        results = es.get(index=index, doc_type=_type, id=_id,
+                         _source=fields)
+    except elasticsearch.exceptions.NotFoundError:
+        return False
+    return results['_source']
+
+
+def process(stage, message):
+    ''' Process a message.
+
+    Implementation of :py:meth:`.AbstractProcessorStage.process` for hooking
+    the stage into DKB workflow.
+
+    :param stage: stage instance
+    :type stage: pyDKB.dataflow.stage.ProcessorStage
+    :param msg: input message with document info
+    :type msg: pyDKB.dataflow.Message
+    '''
+    data = message.content()
+    if type(data) is not dict:
+        log('Incorrect data:' + str(data), 'INPUT')
+        return False
+    _id = data.pop('_id')
+    _type = data.pop('_type')
+    if _id is None or _type is None:
+        log('Insufficient ES info in data:' + str(data), 'INPUT')
+        return False
+
+    # Crutch. Remove unwanted (for now) field added by Stage 016.
+    if 'phys_category' in data:
+        del data['phys_category']
+
+    # Do not check empty documents with valid _id and _type.
+    # It's unlikely that such documents will be produced in DKB. In general,
+    # such documents should be checked by es.exists(), and not es.get().
+    if not data:
+        log('Nothing to check for document (%s, %d)' % (_type, _id), 'INPUT')
+        return False
+
+    es_data = get_fields(INDEX, _id, _type, data.keys())
+    if data != es_data:
+        log('Document (%s, %d) differs between Oracle and ES: Oracle:%s ES:%s'
+            % (_type, _id, data, es_data), 'DIFF')
+        global FOUND_DIFF
+        FOUND_DIFF.append((_type, _id))
+    else:
+        log('Document (%s, %d) is up to date in ES' % (_type, _id), 'INFO')
+
+    return True
+
+
+def main(args):
+    ''' Parse command line arguments and run the stage.
+
+    :param argv: arguments
+    :type argv: list
+    '''
+
+    stage = JSONProcessorStage()
+    stage.add_argument('--conf', help='elasticsearch config', required=True)
+
+    exit_code = 0
+    exc_info = None
+    try:
+        stage.parse_args(args)
+        cfg = load_config(stage.ARGS.conf)
+        stage.process = process
+        es_connect(cfg)
+        if not es.indices.exists(INDEX):
+            log('No such index: %s' % INDEX, 'ERROR')
+            exit_code = 4
+        else:
+            stage.run()
+    except (DataflowException, RuntimeError), err:
+        if str(err):
+            log(err, 'ERROR')
+        exit_code = 2
+    except Exception:
+        exc_info = sys.exc_info()
+        exit_code = 3
+    finally:
+        stage.stop()
+
+    if exc_info:
+        trace = traceback.format_exception(*exc_info)
+        for line in trace:
+            log(line, 'ERROR')
+
+    if exit_code == 0 and FOUND_DIFF:
+        exit_code = 1
+
+    print FOUND_DIFF
```

Comment:
> And the most interesting one: why did you decide to do it this way?

Hmm... I had my doubts about this (I mean the whole way I've chosen to produce the list of inconsistent records, not just these lines), but now, when you mention it, I'm starting to wonder why did I even try to invent something when it is possible to simply put the `(_type, _id)` into output message instead of global variable. This way, at current point the list of records would be obtained by running `data4es-consistency-check` (with `/dev/null` replaced by something). And in the future, when we'll have something to process the inconsistent records with, the script will relate such records to this "something" as soon as it comes across each of them, as an ordinary Processor does (except the fact that it would generate no output messages for consistent records).

Why I decided to do that... I was probably too focused on the fact that this script will be last in the dataflow, and it would be convenient to get the results as batch. My bad, will fix.

While it's already clear that the code should be changed, I'll answer the other questions:

1. > What happens if we run full check -- starting from 2014 and till now? I see it is supposed that we souldn`t have too many "broken" records, so it won`t end up with hundreds of thousands of records in memory; still one won`t see any result (unless he/she is monitoring the log messages) until the whole process is finished.

You pretty much said everything - the broken records will be accumulating in `FOUND_DIFF`, and in theory it is possible to end up with too much information in the memory due to this. Also, the variable won't be displayed/written down/etc until the end of the process. If something goes wrong, it could be lost.

2. > What happens if you run the script in file mode: ./consistency.py -m f ./input/sample2018? And what should happen according to the ./consistency.py -h? OK, maybe in this version of pyDKB it is not that clear, yet I believe you must have seen that PR, #171, aimed in the readability improvement of the help message.

The variable will be printed on screen, while it should be written into file.

3. > What happens if everything works under the supervising of some another application (Kafka-related, in our case)? If everything works fine, it is not supposed to reach this line before the supervising application is stopped manually; and even then I am not sure if it will have any effect (due to the issue mentioned in #45 and #60 -- when Java app stops our subprocess, it seems to make it a bit too forcefully, like saying : "And now you should already be dead.")

It will work according to the supervisor's behavior. This may lead to issues, such as the one you are giving with Java. Supervisors may vary, but the main idea here, I believe, is the same fact that inconsistent records shouldn't be artificially batched together in such way.

Reply:
Done, cacba11.